### PR TITLE
mount custom cache folder using mount providers

### PR DESCRIPTION
**Pull request by icewind1991**
The easy part of https://github.com/owncloud/core/pull/20753

A bit of filesystem cleanup

cc @PVince81 @rullzer please review

### CODE REVIEW

**Comment by rullzer** [comment]
Indeed the easy part.
:+1:

File: lib/private/files/mount/cachemountprovider.php [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+	 * @param IUser $user
+	 * @param IStorageFactory $loader
+	 * @return \OCP\Files\Mount\IMountPoint[]
+	 */
+	public function getMountsForUser(IUser $user, IStorageFactory $loader) {
+		$cacheBaseDir = $this->config->getSystemValue('cache_path', '');
+		if ($cacheBaseDir !== '') {
+			$cacheDir = rtrim($cacheBaseDir, '/') . '/' . $user->getUID();
+			if (!file_exists($cacheDir)) {
+				mkdir($cacheDir, 0770, true);
```
Comment (MorrisJobke) [inline lib/private/files/mount/cachemountprovider.php]: cc @LukasReschke for the scrutinizer comment above ;)

Comment (nickvergessen) [inline lib/private/files/mount/cachemountprovider.php]: Usage of the system config didn't change, so should be fine

File: lib/private/files/mount/cachemountprovider.php [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+namespace OC\Files\Mount;
+
+use OCP\Files\Config\IMountProvider;
+use OCP\Files\Storage\IStorageFactory;
+use OCP\IConfig;
+use OCP\IUser;
+
+/**
+ * Mount provider for custom cache storages
```
Comment (PVince81) [inline lib/private/files/mount/cachemountprovider.php]: What's a "custom cache storage" ?
Is it a custom storage implementation that also comes bundles with its own cache implementation ?

File: lib/private/files/mount/cachemountprovider.php (hunk 1 above)
Comment (PVince81) [inline lib/private/files/mount/cachemountprovider.php]: Looks like this was copy-pasted from above

File: lib/private/files/mount/cachemountprovider.php (hunk 2 above)
Comment (PVince81) [inline lib/private/files/mount/cachemountprovider.php]: Ah, or is this about the "cache" folder, not a storage's "cache" ? (so much confusion, so many caches, better be explicit everywhere nowadays)

File: lib/private/files/mount/cachemountprovider.php (hunk 1 above)
Comment (MorrisJobke) [inline lib/private/files/mount/cachemountprovider.php]: > `$cacheDir` can contain request data and is used in file manipulation context(s) leading to a potential security vulnerability.

Comment (LukasReschke) [inline lib/private/files/mount/cachemountprovider.php]: It's fine. Scrutinizer gets a bit overprotective sometimes :)

**Comment by PVince81** [comment]
Makes sense :+1:

**Comment by PVince81** [comment]
@icewind1991 please resolve conflicts

**Comment by DeepDiver1975** [comment]
rebased - had conflicts

**Pull request merged**